### DISCUSSION TITLE
Implementing repo storage into custom PDS

### DISCUSSION THREAD

**Discussion by rekkih** [bluesky-social/atproto#1973] [How to do it?]
Hey! I've created an app on atproto, but now creating my own PDS (also with typescript), *mostly* for experimenting with the idea of using Supabase as the db and having it live in a worker. I know of `@atproto/repo`, I *really* don't want to implement something like that from scratch, but I don't understand how it handles db interactions, specifically what's expected of ` RepoStorage`. Is this something I have to build myself? If so, where's the best place to start?

I've had a look around in `@atproto/pds` but don't understand it's structure too well.

**Comment by rekkih** [comment]
lol, well, answered myself in a way?

Got a barebones reference [here](https://github.com/bluesky-social/atproto/blob/80161e3c62f8c6c44bf1400b137727bd61c6f350/packages/repo/src/storage/memory-blockstore.ts) (Memory blockstore).

**Comment by DavidBuchanan314** [comment]
I'm not familiar enough with the typescript codebase to answer your question directly, but implementing it from scratch *might* be more approachable than you'd think. Even if you don't go that route, you might find these useful for reference:

https://github.com/snarfed/arroba - A from-scratch Python PDS implementation

https://github.com/DavidBuchanan314/picopds - Also a from-scratch Python PDS implementation, but a bit less "production ready", which also means there's less code to understand.

I know neither of these are typescript, but they're both smaller codebases, and perhaps that makes it easier to see what's going on.

**Reply by rekkih** [reply]
I did have a look at arroba before, and it's repo storage references the typescript implementation. Looking at how the repo works more I was just looking in the wrong places (being, the PDS' implementation). I was misunderstanding what the repo was actually doing.

You might be right though, implementing from scratch might honestly be better, especially to learn from. Thanks!
